File: Leetcode/0133 Clone Graph.ts
function cloneGraph(node: Node | null): Node | null {
    if(node == null) {
        return null;
    }
    return clone(node, new Map());
}

function clone(node: Node, alreadyClonned: Map<number, Node>) {
    if(alreadyClonned.has(node.val)) {
        return alreadyClonned.get(node.val);
    }
    const clonnedNode = new Node(node.val, []);
    alreadyClonned.set(clonnedNode.val, clonnedNode);
    for(const neighbor of node.neighbors) {
        clonnedNode.neighbors.push(clone(neighbor, alreadyClonned));
    }
    return clonnedNode;
}

File: Leetcode/0133 Clone Graph.py
class Solution:
    def cloneGraph(self, node):
        if node is None:
            return None
        return dfs(node, {})


def dfs(node, already_clonned):
    if node.val in already_clonned:
        return already_clonned[node.val]
    clonned = Node(node.val, [])
    already_clonned[clonned.val] = clonned
    for neighbor in node.neighbors:
        clonned.neighbors.append(dfs(neighbor, already_clonned))
    return clonned
